include/kernels/dslash_misc.cuh

```diff
@@ -0,0 +1,314 @@
+#pragma once
+
+#include <color_spinor_field_order.h>
+#include <clover_field_order.h>
+#include <index_helper.cuh>
+#include <color_spinor.h>
+#include <linalg.cuh>
+#include <kernel.h>
+
+namespace quda {
+
+  /**
+     @brief Parameter structure for driving init_dslash_atomic
+  */
+  template <typename T_> struct init_dslash_atomic_arg : kernel_param<> {
+    using T = T_;
+    T *count;
+
+    init_dslash_atomic_arg(T *count, unsigned int size) :
+      kernel_param(dim3(size, 1, 1)),
+      count(count) { }
+  };
+
+  /**
+     @brief Functor that uses placement new constructor to initialize
+     the atomic counters
+  */
+  template <typename Arg> struct init_dslash_atomic {
+    const Arg &arg;
+    static constexpr const char *filename() { return KERNEL_FILE; }
+    constexpr init_dslash_atomic(const Arg &arg) : arg(arg) { }
+    __device__ void operator()(int i) { new (arg.count + i) typename Arg::T {0}; }
+  };
+
+  /**
+     @brief Parameter structure for driving init_dslash_arr
+  */
+  template <typename T_> struct init_arr_arg : kernel_param<> {
+    using T = T_;
+    T *arr;
+    T val;
+
+    init_arr_arg(T *arr, T val, unsigned int size) :
+      kernel_param(dim3(size, 1, 1)),
+      arr(arr),
+      val(val) { }
+  };
+
+  /**
+     @brief Functor to initialize the arrive signal
+  */
+  template <typename Arg> struct init_sync_arr {
+    const Arg &arg;
+    static constexpr const char *filename() { return KERNEL_FILE; }
+    constexpr init_sync_arr(const Arg &arg) : arg(arg) { }
+    __device__ void operator()(int i) { *(arg.arr + i) = arg.val; }
+  };
+
+  /**
+     @brief Parameter structure for driving the Gamma operator
+   */
+  template <typename Float, int nColor_>
+  struct GammaArg : kernel_param<> {
+    using real = typename mapper<Float>::type;
+    constexpr static int nColor = nColor_;
+    typedef typename colorspinor_mapper<Float,4,nColor>::type F;
+
+    F out;                // output vector field
+    const F in;           // input vector field
+    const int d;          // which gamma matrix are we applying
```

Comment:
is d just nexr after a,b,c or is there a good reason tu use set to select the gamma matrix?